---
title: "Using a VariantExperiment data object to manage COVID19 data"
author: "Jerome Ambroise, Laurent Gatto, Julie Hurel, Bertrand Bearzatto, Jean-Luc Gala"
date: '`r format(Sys.time(), "%B %d, %Y,%H:%M")`'
output:
  html_document:
    smart: FALSE
    code_folding: show
    collapsed: yes
    fig_caption: yes
    fig_height: 6
    fig_width: 9
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 3
    toc_float: yes
editor_options:
  chunk_output_type: console
---

In this document we illustrate the advantages of using the
`r BiocStyle::Biocpkg("VariantExperiment")` Bioconductor
package for handling SARS-CoV-2 genomic data and associated
metadata. For more details, we invite you to read the associated
publication ('On the many advantages of using the VariantExperiment
Bioconductor package for storing, exchanging and analyzing SARS-CoV-2
genomic data and associated metadata' available on bioRxiv.org).


# Load package

```{r, message = FALSE}
library(VariantExperiment)
```

# Data loading

The `RangedSummarizedExperiment` object was created with the
VariantExperiment package for handling SARS-COV2 genomic data and
associated 'dummy' (in the present study) metadata.

```{r}
sarscov2_ve <- readRDS('sarscov2_ve.rds')
```

# Data management

The object contains `r nrow(sarscov2_ve)` variants sequenced in
`r ncol(sarscov2_ve)` different samples.

```{r}
sarscov2_ve
```

## `Assay`: genomic data

As shown heren, the assays slots store the Genomic data (including the
called genotype and a quality metrics such as read depth (DP) for the
illustration).  Of course other quality metrics could be added as a
phred score.

```{r, warning = FALSE}
assay(sarscov2_ve, "genotype")
assay(sarscov2_ve, "DP")
```

It now becomes easy to produce a graphical overview of the genotypes
of the read depth of a dataset by plotting the respective assay
heatmap.

```{r, fig.cap = "Heatmap of the virus genotypes."}
pheatmap::pheatmap(assay(sarscov2_ve, "genotype"))
```

```{r, fig.cap = "Heatmap of the sequencing read depth illustrating how some samples have systematic low read depth (blue)."}
pheatmap::pheatmap(assay(sarscov2_ve, "DP"))
```


## `colData`: sample annotations/metadata

As shown here, the `colData` slot stores the annotation/metadata
associated to each sample (sample, pre-analytic, analytic,
bioinformatics)

```{r}
SummarizedExperiment::colData(sarscov2_ve)
```

## `rowData`: feature annotationnns

Finally, the `rowData` includes the description of each mutation. This
slot could also include an indicator reflecting if this variant is
currently considered as a Variant of Concern (VOC) or Variant of high
Concern.

```{r}
SummarizedExperiment::rowRanges(sarscov2_ve)
```

# Subsetting

Here we illustrate how to subset the object. For the demonstration we
select only data:

- from patient with obesity (Obesity = yes)
- mutation with a start position < 5000

```{r}
sel_start <- start(SummarizedExperiment::rowRanges(sarscov2_ve)) < 5000
sel_obesity <- sarscov2_ve$Obesity == 'yes'

ve2 <- sarscov2_ve[sel_start, sel_obesity]
ve2
```

As shown here the selection is performed in one operation for both the
rowData, colData and assay(s). This property therefore ensures to keep
a perfect synchronization of genomic data with metadata.

We can also make use the the VOC/VOI definition in the rowData slot to
extract and focus on the set of mutated loci of each variant:

```{r}
voc1 <- sarscov2_ve[rowData(sarscov2_ve)$VOC1, ]
assay(voc1, "genotype")
rowRanges(voc1)
```

```{r}
voc2 <- sarscov2_ve[rowData(sarscov2_ve)$VOC2, ]
assay(voc2, "genotype")
rowRanges(voc2)
```

Below, we use the sample-level metadata to extract the samples that
have been annotated as having been infected by certain variants of
concern:

```{r}
has_VOC1 <- sarscov2_ve[, sarscov2_ve$has_VOC1]
assay(has_VOC1, "genotype")
```
```{r}
has_VOC2 <- sarscov2_ve[, sarscov2_ve$has_VOC2]
assay(has_VOC2, "genotype")
```

Both subsetting can of course be applied simultaneously to extract the
variants of concern loci in the patients that have been infected by that particular VOC:

```{r}
is_voc1 <- sarscov2_ve[rowData(sarscov2_ve)$VOC1, sarscov2_ve$has_VOC1]
assay(is_voc1, "genotype")
```


```{r}
is_voc2 <- sarscov2_ve[rowData(sarscov2_ve)$VOC2, sarscov2_ve$has_VOC2]
assay(is_voc2, "genotype")
```

# Creation of the VariantExperiment object

1. Loading required packages

```{r, message = FALSE}
library(SummarizedExperiment)
library(VariantExperiment)
library(VariantAnnotation)
library(SeqArray)
```

2. Creation of the object based on a vcf file

```{r}
vcf.header <- seqVCF_Header('variant.q20.vcf')
seqVCF2GDS(vcf.fn = 'variant.q20.vcf',
           out.fn = 'sarscov2.gds',
           info.import = c('GT','DP'))
sarscov2_ve <- makeVariantExperimentFromGDS('sarscov2.gds')
assayNames(sarscov2_ve) <- c('genotype','DP')
sarscov2_ve
```

3. Sample annotations/metadata (aka `colData`)

```{r}
col_data <- DataFrame(read.csv('coldata.csv'))
rownames(col_data) <- col_data$Sample.ID
col_data <- col_data[match(colnames(sarscov2_ve), col_data$SRA.ID), ]
stopifnot(identical(colnames(sarscov2_ve), col_data$SRA.ID))
SummarizedExperiment::colData(sarscov2_ve) <- col_data
```

4. Feature annotations/medata (aka `rowData`)


```{r}
vcffile.fb <- readVcf('variant.q20.vcf')
rowRanges(sarscov2_ve) <- SummarizedExperiment::rowRanges(vcffile.fb)
```

5. Annotation of variants of interest/concern

```{r}
## VOC initialisation
rowData(sarscov2_ve)$VOC1 <- rowData(sarscov2_ve)$VOC2 <- FALSE
## Definition of VOC1, defined by 2 loci, namely
## "MN908947.3:5224_TAAAAAG/TAAAAG" and
## "MN908947.3:22487_GAAAAAG/GAAAAG"
rowData(sarscov2_ve)$VOC1[c(7, 26)] <- TRUE
## Definition of VOC2, defined by only
## "MN908947.3:29377_TAAAAAG/TAAAAG"
rowData(sarscov2_ve)$VOC2[38] <- TRUE
```

```{r}
## has_VOC initialisation
sarscov2_ve$has_VOC1 <- sarscov2_ve$has_VOC2 <- FALSE
## Annotation of samples infected by VOC1 and VOC2
sarscov2_ve$has_VOC1[c(2, 6, 9)] <- TRUE
sarscov2_ve$has_VOC2[c(3, 5, 8)] <- TRUE
```



6. Serialisation of the object

```{r}
saveRDS(sarscov2_ve, 'sarscov2_ve.rds')
```

# Further reading

The `VariantExperiment` package vignettes

- [VariantExperiment: A RangedSummarizedExperiment Container for VCF/GDS Data with GDS Backend](http://bioconductor.org/packages/release/bioc/vignettes/VariantExperiment/inst/doc/VariantExperiment-class.html)
- [VariantExperiment methods](http://bioconductor.org/packages/release/bioc/vignettes/VariantExperiment/inst/doc/VariantExperiment-methods.html)

# Session info

```{r}
sessionInfo()
```
