---
title: "Demonstration of using a VariantExperiment file to manage COVID19 data"
author: "Jerome Ambroise, Laurent Gatto, Julie Hurel, Bertrand Bearzatto, Jean-Luc Gala"
date: '`r format(Sys.time(), "%B %d, %Y,%H:%M")`'
output: 
  html_document:
    smart: FALSE
    code_folding: show
    collapsed: yes
    fig_caption: yes
    fig_height: 6
    fig_width: 9
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 3
    toc_float: yes
editor_options: 
  chunk_output_type: console
---

In this document we illustrate the advantages of using the VariantExperiment Bioconductor package for handling SARS-CoV-2 genomic data and associated metadata. For more details, we invite you to read the associated publication ('On the many advantages of using the VariantExperiment Bioconductor package for storing, exchanging and analyzing SARS-CoV-2 genomic data and associated metadata' available on bioRxiv.org).


# Load package

```{r message=FALSE}
library(VariantExperiment)
```

# Load file

THe RangedSummarizedExperiment object was created with the VariantExperiment package for handling SARS-COV2 genomic data and associated 'dummy' (in the present study) metadata.

```{r}
SARSCOV2 <- readRDS( 'SARSCOV2_VariantExperiment.RDS')
```

# Explore the file

```{r}
SARSCOV2
```

## Assay data  = Genomic data 

As shown heren, the assays slots store the Genomic data (including the called genotype and a quality metrics [e.g.: Read Depth DP for the illustration]).
Of course other quality metrics could be added as a phred score.

```{r}
assays(SARSCOV2)$genotype
assays(SARSCOV2)$DP
```

## colData  = Metadata associated to each sample

As shown here, the colData slots store the metadata associated to each sample (sample, pre-analytic, analytic, bioinformatics)

```{r}
colData(SARSCOV2)
```

## rowData  = feature data 

Finally, the rowData includes the description of each mutation. This slots could also include an indicator reflecting if this variant is currently considered as a Variant of Concern (VOC) or Variant of high Concern.

```{r}
rowData(SARSCOV2)
```


# Subsetting

Here we illustrate how to subset the object. For the demonstration we select only data:
- from patient with obesity (Obsesity = yes)
- mutation with a start position < 5000 

```{r}
COVID.FILE.Subset <- SARSCOV2[rowData(SARSCOV2)$start<5000,colData(SARSCOV2)$Obesity=='yes']
```

As shown here the selection is performed in one operation for both the rowData, colData and assay(s). This property therefore ensures to keep a perfect synchronization of genomic data with metadata.


# Session info

```{r}
sessionInfo()
```



